### DISCUSSION TITLE
Flashing troubles

### DISCUSSION THREAD

**Discussion by dbrgn** [gcormier/megadesk#49] [General]
I finished building a v5 PCB 🙂🎉 

![image](https://user-images.githubusercontent.com/105168/108135976-66bf3600-70b9-11eb-967f-0d5d91053dc3.png)

Unfortunately flashing the ATtiny doesn't seem to work:

```
$ avrdude -c usbasp -p t841 -U lfuse:w:0xe2:m

avrdude: error: program enable: target doesn't answer. 1
avrdude: initialization failed, rc=-1
         Double check connections and try again, or use -F to override
         this check.
```

I used an USBasp programmer (with 5V output) and the current SVN version of avrdude (because the latest release doesn't support the t841) to flash the megadesk PCB. The PCB was only connected to the programmer, not to the desk.

I measured the voltage across the 5V and GND pins on the programming header, and it shows 4V. Without the megadesk PCB connected, the voltage provided by USBasp is 5V. Could be some kind of short, but the components don't get warm...

Unfortunately the schematics in KiCad look all garbled to me, I probably have the wrong symbols? (Incredible, that in 2021 this still doesn't work properly.) Should the 5V power supply from the PCB suffice, or is the 24V power input from the table required as well, for in-circuit programming?

**Comment by gcormier** [comment]
For ICSP, all that is required is the 5V.

Is it the lighting, or is there a chunk missing from your attiny841?! :)

**Comment by dbrgn** [comment]
Ooof, I think I got D1 soldered in the wrong way around. Those little bastards are really hard to read properly.

Do you have a PDF of the v5 schematics, so I can check what I could possibly have destroyed with that?

> Is it the lighting, or is there a chunk missing from your attiny841?! :)

Just the lighting, due to some flux residues 😂

**Comment by gcormier** [comment]
Sure here you go
![image](https://user-images.githubusercontent.com/8388983/108135933-01525180-7087-11eb-9dcb-45bbf95fb12c.png)

**Comment by gcormier** [comment]
Which programmer are you using? At least you're on linux.

`avrdude` sure has given me some "fun" times.

**Comment by gcormier** [comment]
Note for V5, if you are using an IDC connector and the typical AVR programmers, the notch on `v5` points up.

On v1 it pointed down
![image](https://user-images.githubusercontent.com/8388983/108136155-76be2200-7087-11eb-8219-5394dfcc6861.png)

**Comment by dbrgn** [comment]
I'm using an usbasp variant: https://www.fischl.de/usbasp/ I've used in the past for Arduinos and ATtinys and it worked fine so far.

Ok, so if I have D1 reversed (which I checked with a multimeter) then it connects 5V to 12V, which feeds backwards into the regulator. Probably nothing that should destory the regulator, but I'll double-check after fixing it.

> Note for V5, if you are using an IDC connector and the typical AVR programmers, the notch on v5 points up.

Thanks for the hint. I used jumper wires for now, since my programmer has a 10-pin connector 🙂 I think the reversed diode should be the reason, dropping the supply voltage below the 5V required to run the ATtiny.

I'll fix this another day. Thanks already for your quick help!

**Comment by gcormier** [comment]
Great, let me know how it goes!

You can do all the programming, and you should hear the startup beeps with merely the 5v/programmer connected. No need to connect the desk to hear the beeps. Once you hit that point you should be good to switch over.

**Comment by dbrgn** [comment]
> You can do all the programming, and you should hear the startup beeps with merely the 5v/programmer connected. No need to connect the desk to hear the beeps.

Ah, very helpful to know! Thanks!

**Comment by dbrgn** [comment]
Hm, I fixed the diode, but as soon as I connect the 5V input to the ICSP header, it drops to 4V again.

Do you see potential for a component to get destroyed by the reversed D1? I'm a bit stuck with debugging. The diodes are still fine, according to my multimeter. I guess as a next step I'd probably desolder U2 to see if the problem still happens that way.

**Comment by gcormier** [comment]
Do you have another attiny841?
Do you have a bench PSU capable of measuring current? Or can you hook up some power to it with a meter to measure the current? It should be under 30mA at all times.

**Comment by dbrgn** [comment]
I finally got around to debugging this. I connected the PCB to the bench PSU and hooked up a multimeter. It drew 22 mA and kept the voltage steady at 5V.

Then I connected the UsbASP programmer again, and voltage dropped back to 4V.

I think the issue is that either the programmer is broken, or that it cannot deliver 22 mA. But that would be weird (I think the 5V output comes straight from USB), so it's probably just broken. I'll order a new programmer and will let you know how that goes!

**Reply by dbrgn** [reply]
I received an USBtinyISP today and tried to flash, but without luck. Voltage looks better, but initialization in avrdude (connecting to the MCU) fails.

I think next I'll try to hook up a new attiny841 directly to the programmer to see if flashing works. If it does, I'll desolder the one on megadesk and will retry with that. Maybe I somehow damaged it.

**Comment by dbrgn** [comment]
Huh, I'm beginning to doubt my sanity 🙈 I took a completely new ATTINY841-SSUR and soldered it to a SOIC-14 breakout. I then connected the pins from my USBtinyISP board to the breakout:

- VCC - Pin 1
- GND - Pin 14
- nRST - Pin 4 (PB3)
- MOSI - Pin 7 (PA6)
- MISO - Pin 8 (PA5)
- SCK - Pin 9 (PA4)

I then ran `avrdude -c usbtiny -p attiny841`. But I still keep tetting the `avrdude: initialization failed, rc=-1` error 😕

@gcormier what avrdude version do you use? I'm on 6.3-20171130. (The version is printed with `avrdude --help`)

**Reply by dbrgn** [reply]
I even hooked up a logic analyzer to the programmer while programming. Doesn't look too wrong (but I don't really know how the ISP SPI protocol is supposed to work).

![2021-03-21-005202_1488x626_scrot](https://user-images.githubusercontent.com/105168/111888766-bad27880-89df-11eb-874d-c7ba893a73a0.png)

**Reply by dbrgn** [reply]
I did another try with Arduino IDE + ATTinyCore, and that seemed to do the trick.

I'm starting to wondering whether the "error message" was just bad UX when not passing in a hex file to be flashed 😕 Maybe it's also a bug in avrdude, I'm not sure. Setting the fuses doesn't work either with avrdude.

Anyways, it seems that flashing works with Arduino IDE + ATTinyCore, so I guess I'll stick with that for now (or maybe PlatformIO). Sorry for bugging you with these issues 🙂 But I was a bit confused since I've never had an issue with avrdude in the past when working with the ATtiny13a.

Edit: It only works when choosing "USBtinyISP (ATTinyCore) SLOW, for new or 1 MHz parts" as programmer, not with "USBtinyISP (ATTinyCore) FAST, for parts running >=2 MHz". It seems that with this config ATTinyCore sets the [-B8 parameter](https://github.com/SpenceKonde/ATTinyCore/blob/bdafb491ff107e0309410e32a2c6c3882bb31718/avr/programmers.txt#L24-L28) for avrdude, but invoking avrdude manually with that option (or even slower speeds) does not help for some reason.

Edit 2: After setting the fuses using the "burn bootloader" option in the Arduino IDE in the SLOW mode, uploading the code also works in FAST mode.

**Comment by gcormier** [comment]
No problem! If I had a dollar for every hour I chased my own tail with
avrdude I'd be rich!

I ran into the same issue with a pololu programmer and the initial
programming speed. Glad to hear you got it sorted out!
